tests/test_kv_store.c

```diff
@@ -60,11 +102,77 @@ static void test_type_H(void **state)
 	}
 }
 
+static void test_kvstore_iterate(void **state)
+{
+	struct iovec           test_iov[_KV_VALUE_IDX_COUNT];
+	size_t                 data_size, kv_size;
+	struct iovec *         return_iov;
+	kv_store_iter_t *      iter;
+	kv_store_value_flags_t flags        = KV_STORE_VALUE_REF | KV_STORE_VALUE_VECTOR;
+	uint64_t               seqnum       = 0;
+	sid_resource_t *       kv_store_res = NULL;
+	struct kv_update_arg   update_arg;
+	size_t                 meta_size = 0;
+
+	kv_store_res = sid_resource_create(SID_RESOURCE_NO_PARENT,
+	                                   &sid_resource_type_kv_store,
+	                                   SID_RESOURCE_RESTRICT_WALK_UP,
+	                                   "testkvstore",
+	                                   &main_kv_store_res_params,
+	                                   SID_RESOURCE_PRIO_NORMAL,
+	                                   SID_RESOURCE_NO_SERVICE_LINKS);
+
+	KV_VALUE_PREPARE_HEADER(test_iov, seqnum, flags, (char *) TEST_OWNER);
+	test_iov[KV_VALUE_IDX_DATA].iov_base = "test";
+	test_iov[KV_VALUE_IDX_DATA].iov_len  = sizeof("test");
+
+	update_arg = (struct kv_update_arg) {.res      = kv_store_res,
+	                                     .gen_buf  = NULL,
+	                                     .owner    = TEST_OWNER,
+	                                     .custom   = NULL,
+	                                     .ret_code = -EREMOTEIO};
+
+	/* Add the whole vector with TEST_KEY as the key */
+	assert_ptr_not_equal(kv_store_set_value(kv_store_res,
+	                                        TEST_KEY,
+	                                        test_iov,
+	                                        KV_VALUE_IDX_DATA + 1,
+	                                        KV_STORE_VALUE_VECTOR,
+	                                        KV_STORE_VALUE_NO_OP,
+	                                        _kv_overwrite,
+	                                        &update_arg),
+	                     NULL);
+
+	assert_ptr_not_equal(iter = kv_store_iter_create(kv_store_res), NULL);
+	/* validate the contents of the kv store */
+	while ((return_iov = kv_store_iter_next(iter, &data_size, &flags))) {
+		assert_true(flags & (KV_STORE_VALUE_REF | KV_STORE_VALUE_VECTOR));
```

Comment:
...so, if you called `kv_store_set_value(kv_store_res, ..., KV_STORE_VALUE_VECTOR, ...)`, the flags tested should be only `KV_STORE_VALUE_VECTOR` (without `KV_STORE_VALUE_REF`).

This test line passes now, because `flags & (KV_STORE_VALUE_REF | KV_STORE_VALUE_VECTOR` is true if either `KV_STORE_VALUE_REF` is set or `KV_STORE_VALUE_VECTOR` is set or both. So we need to check the exact flags here with `==` instead of `&` I think (and without `KV_STORE_VALUE_REF` if we're not using it for `kv_store_set_value`)